libs/utils/trace.py

```diff
@@ -728,6 +734,77 @@ def _loadFunctionsStats(self, path='trace.stats'):
 
         return len(self._functions_stats_df) > 0
 
+    @memoized
+    def getCPUActiveSignal(self, cpu):
+        """
+        Build a square wave representing the active (i.e. non-idle) CPU time,
+        i.e.:
+            cpu_active[t] == 1 if at least one CPU is reported to be
+                               non-idle by CPUFreq at time t
+            cpu_active[t] == 0 otherwise
+
+        :param cpu: CPU ID
+        :type cpu: int
+
+        :returns: :mod:`pandas.Series`
+        """
+        if not self.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'cannot compute CPU active signal!')
+            return None
+
+        idle_df = self._dfg_trace_event('cpu_idle')
+        cpu_df = idle_df[idle_df.cpu_id == cpu]
+
+        cpu_active = cpu_df.state.apply(
+            lambda s: 1 if s == NON_IDLE_STATE else 0
+        )
+
+        start_time = 0.0
+        if not self.ftrace.normalized_time:
+            start_time = self.ftrace.basetime
+        if cpu_active.index[0] != start_time:
+            entry_0 = pd.Series(cpu_active.iloc[0] ^ 1, index=[start_time])
+            cpu_active = pd.concat([entry_0, cpu_active])
+
+        return cpu_active
+
+    @memoized
+    def getClusterActiveSignal(self, cluster):
+        """
+        Build a square wave representing the active (i.e. non-idle) cluster
+        time, i.e.:
+            cluster_active[t] == 1 if at least one CPU is reported to be
+                                   non-idle by CPUFreq at time t
+            cluster_active[t] == 0 otherwise
+
+        :param cluster: list of CPU IDs belonging to a cluster
+        :type cluster: list(int)
+
+        :returns: :mod:`pandas.Series`
+        """
+        if not self.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'cannot compute cluster active signal!')
+            return None
+
+        cpu_active = {}
+        for cpu in cluster:
+            cpu_active[cpu] = self.getCPUActiveSignal(cpu)
+
+        active = pd.DataFrame(cpu_active)
```

Comment:
It seems to have problems:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-56aca3433a47> in <module>()
----> 1 trace.data_frame.idle_state_residency('big')

/media/hdd1tb/work/lisa/libs/utils/analysis/idle_analysis.pyc in _dfg_idle_state_residency(self, entity)
     76         #                       idle by cpufreq at time t
     77         #     entity_idle[t] == 0 otherwise
---> 78         entity_idle = self._trace.getClusterActiveSignal(_entity) ^ 1
     79 
     80         # In order to compute the time spent in each idle statefrequency we

/media/hdd1tb/work/lisa/libs/devlib/devlib/utils/misc.pyc in memoize_wrapper(*args, **kwargs)
    546                               for k, v in kwargs.iteritems())
    547         if id_string not in __memo_cache:
--> 548             __memo_cache[id_string] = func(*args, **kwargs)
    549         return __memo_cache[id_string]
    550 

/media/hdd1tb/work/lisa/libs/utils/trace.pyc in getClusterActiveSignal(self, cluster)
    793             cpu_active[cpu] = self.getCPUActiveSignal(cpu)
    794 
--> 795         active = pd.DataFrame(cpu_active)
    796         active.fillna(method='ffill', inplace=True)
    797 

/usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in __init__(self, data, index, columns, dtype, copy)
    224                                  dtype=dtype, copy=copy)
    225         elif isinstance(data, dict):
--> 226             mgr = self._init_dict(data, index, columns, dtype=dtype)
    227         elif isinstance(data, ma.MaskedArray):
    228             import numpy.ma.mrecords as mrecords

/usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in _init_dict(self, data, index, columns, dtype)
    361 
    362         return _arrays_to_mgr(arrays, data_names, index, columns,
--> 363                               dtype=dtype)
    364 
    365     def _init_ndarray(self, values, index, columns, dtype=None,

/usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in _arrays_to_mgr(arrays, arr_names, index, columns, dtype)
   5161 
   5162     # don't force copy because getting jammed in an ndarray anyway
-> 5163     arrays = _homogenize(arrays, index, dtype)
   5164 
   5165     # from BlockManager perspective

/usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in _homogenize(data, index, dtype)
   5463                 # Forces alignment. No need to copy data since we
   5464                 # are putting it into an ndarray later
-> 5465                 v = v.reindex(index, copy=False)
   5466         else:
   5467             if isinstance(v, dict):

/usr/local/lib/python2.7/dist-packages/pandas/core/series.pyc in reindex(self, index, **kwargs)
   2266     @Appender(generic._shared_docs['reindex'] % _shared_doc_kwargs)
   2267     def reindex(self, index=None, **kwargs):
-> 2268         return super(Series, self).reindex(index=index, **kwargs)
   2269 
   2270     @Appender(generic._shared_docs['fillna'] % _shared_doc_kwargs)

/usr/local/lib/python2.7/dist-packages/pandas/core/generic.pyc in reindex(self, *args, **kwargs)
   1960         # perform the reindex on the axes
   1961         return self._reindex_axes(axes, level, limit, tolerance,
-> 1962                                   method, fill_value, copy).__finalize__(self)
   1963 
   1964     def _reindex_axes(self, axes, level, limit, tolerance, method,

/usr/local/lib/python2.7/dist-packages/pandas/core/generic.pyc in _reindex_axes(self, axes, level, limit, tolerance, method, fill_value, copy)
   1979             obj = obj._reindex_with_indexers(
   1980                 {axis: [new_index, indexer]},
-> 1981                 fill_value=fill_value, copy=copy, allow_dups=False)
   1982 
   1983         return obj

/usr/local/lib/python2.7/dist-packages/pandas/core/generic.pyc in _reindex_with_indexers(self, reindexers, fill_value, copy, allow_dups)
   2071                                                 fill_value=fill_value,
   2072                                                 allow_dups=allow_dups,
-> 2073                                                 copy=copy)
   2074 
   2075         if copy and new_data is self._data:

/usr/local/lib/python2.7/dist-packages/pandas/core/internals.pyc in reindex_indexer(self, new_axis, indexer, axis, fill_value, allow_dups, copy)
   3501         # some axes don't allow reindexing with dups
   3502         if not allow_dups:
-> 3503             self.axes[axis]._can_reindex(indexer)
   3504 
   3505         if axis >= self.ndim:

/usr/local/lib/python2.7/dist-packages/pandas/core/index.pyc in _can_reindex(self, indexer)
   2084         # trying to reindex on an axis with duplicates
   2085         if not self.is_unique and len(indexer):
-> 2086             raise ValueError("cannot reindex from a duplicate axis")
   2087 
   2088     def reindex(self, target, method=None, level=None, limit=None,

ValueError: cannot reindex from a duplicate axis
```

the following seems to fix it, but it might be less efficient:
```

-        cpu_active = {}
+        active = None
         for cpu in cluster:
-            cpu_active[cpu] = self.getCPUActiveSignal(cpu)
+            if isinstance(active, pd.DataFrame):
+                active = active.join(pd.DataFrame(self.getCPUActiveSignal(cpu), columns=['{}'.format(cpu)]), how='outer')
+            else:
+                active = pd.DataFrame(self.getCPUActiveSignal(cpu), columns=['{}'.format(cpu)])
 
-        active = pd.DataFrame(cpu_active)
```